### PR TITLE
Metafeeds graph changes

### PR DESCRIPTION
**Pull request by cryptix**
changes required to index metafeed updates on the trust/follow graph package such that they are replicated, too.

(based on #102 - merge first)

### CODE REVIEW

File: sbot/new.go [hunk 1]
```diff
@@ -484,10 +481,41 @@ func New(fopts ...Option) (*Sbot, error) {
...
+			if err != nil {
+				return false
+			}
+
+			if len(bencoded) != 2 {
+				return false
+			}
+
+			var justTheType metamngmt.Typed
+			err = bencode.DecodeBytes(bencoded[0], &justTheType)
```
Comment (cryptix) [inline sbot/new.go:507]: comment on why `[0]`

**Review by cryptix — COMMENTED** [review]
some comments

File: graph/builder.go [hunk 2]
```diff
@@ -48,16 +47,21 @@ type IndexingBuilder interface {
...
 }
 
-type builder struct {
+type BadgerBuilder struct {
 	kv *badger.DB
 
-	idx     librarian.SeqSetterIndex
-	idxSink librarian.SinkIndex
+	idx librarian.SeqSetterIndex
+	// idxMetafeeds librarian.SeqSetterIndex
```
Comment (cblgh) [inline graph/builder.go:54]:
```suggestion
```

File: graph/builder.go [hunk 3]
```diff
@@ -48,16 +47,21 @@ type IndexingBuilder interface {
 	OpenIndex() (librarian.SeqSetterIndex, librarian.SinkIndex)
 }
 
-type builder struct {
+type BadgerBuilder struct {
```
Comment (cblgh) [inline graph/builder.go:50]: rename to BendyBuilder? iirc there's not bipfy badger anymore

File: graph/metafeeds_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/ssb-ngi-pointer/go-metafeed/metakeys"
+	"github.com/ssb-ngi-pointer/go-metafeed/metamngmt"
+	refs "go.mindeco.de/ssb-refs"
+)
+
+var metafeedsScenarios = []PeopleTestCase{
+	{
+		name: "metafeeds, simple",
+		ops: []PeopleOp{
+			PeopleOpNewPeerWithAglo{"alice", refs.RefAlgoFeedBendyButt},
```
Comment (cblgh) [inline graph/metafeeds_test.go:18]: this is still aglo tho :)

`PeopleOpNewPeerWithAglo`

File: graph/metafeeds_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+				of:    "alice",
+				name:  "alice-legacy",
+				nonce: "test1",
+				algo:  refs.RefAlgoFeedSSB1,
+			},
+
+			PeopleOpFollow{"alice-legacy", "some"},
+		},
+		asserts: []PeopleAssertMaker{
+			PeopleAssertHops("alice", 0, "alice-legacy", "some"),
```
Comment (cblgh) [inline graph/metafeeds_test.go:81]: why would the hops from `alice/alice-legacy` to `some` be 0? i would expect 1

File: graph/metafeeds_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		},
+	},
+}
+
+type PeopleOpNewSubFeed struct {
+	of, nonce, name string
+	algo            refs.RefAlgo
+}
+
+func (op PeopleOpNewSubFeed) Op(state *testState) error {
```
Comment (cblgh) [inline graph/metafeeds_test.go:91]: each `Op` command could have an explanatory package comment one-liner. reading the subsequent code would be easier then, i think :)

File: repo/filtered_log.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+// NewFilteredLog wraps the passed log into a new one, using the FilterFunc to decide if a message is in the log.
+func NewFilteredLog(b margaret.Log, fn FilterFunc) margaret.Log {
+	return FilteredLog{
+		backing: b,
+		filter:  fn,
+	}
+}
+
+type FilteredLog struct {
```
Comment (cblgh) [inline repo/filtered_log.go:24]: i believe FilteredLog is an essential concept, can you comment its overarching purpose?

File: repo/filtered_log.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	backing margaret.Log
+
+	filter FilterFunc
+}
+
+func (fl FilteredLog) Seq() luigi.Observable { return fl.backing.Seq() }
+
+// Get retreives the message object by traversing the authors sublog to the root log
```
Comment (cblgh) [inline repo/filtered_log.go:32]:
```suggestion
// Get retrieves the message object by traversing the authors sublog to the root log

File: repo/filtered_log.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+func (fl FilteredLog) Get(s margaret.Seq) (interface{}, error) {
+	v, err := fl.backing.Get(s)
+	if err != nil {
+		return nil, fmt.Errorf("publish get: failed to retreive sequence for the root log: %w", err)
```
Comment (cblgh) [inline repo/filtered_log.go:36]:
```suggestion
		return nil, fmt.Errorf("publish get: failed to retrieve sequence for the root log: %w", err)

File: sbot/metafeed_test.go [hunk 10]
```diff
@@ -184,31 +186,73 @@ func TestMultiFeedManualSync(t *testing.T) {
...
+	_, err = multiBot.MetaFeeds.Publish(subfeedClassic, refs.NewContactFollow(receiveBot.KeyPair.ID()))
+	r.NoError(err)
+	_, err = receiveBot.PublishLog.Publish(refs.NewContactFollow(multiBot.KeyPair.ID()))
+	r.NoError(err)
+
+	// multiBot.Replicate(receiveBot.KeyPair.ID())
+	// receiveBot.Replicate(multiBot.KeyPair.ID())
+	// receiveBot.Replicate(subfeedClassic)
+	// receiveBot.Replicate(subfeedGabby)
+
```
Comment (cblgh) [inline sbot/metafeed_test.go:218]:
```suggestion

**Review by cblgh — COMMENTED** [review]
i left some comments & questions that might benefit the code base from being addressed, but otherwise looks good afaics

File: graph/builder.go (hunk 3 above)
Comment (cryptix) [reply graph/builder.go:50]: nope, will clear this up with a comment.

This is about the [badger key-value db](https://github.com/dgraph-io/badger) that powers the index system.

File: graph/metafeeds_test.go (hunk 5 above)
Comment (cryptix) [reply graph/metafeeds_test.go:81]: 1 would be follows of friends, too but alice and alice legacy are essentially the same person (ie sameAs)

see people_test for the other hop cases.

File: graph/builder.go (hunk 3 above)
Comment (cblgh) [reply graph/builder.go:51]: oh lol

**Pull request merged**